File: online judge.kaner murillo/c++/black list/551.cpp
#include<stack>
#include<cstdio>
#include<iostream>
#include<string>

using namespace std;

int main(){

    int contador, i;
    string cadena, nomb, letra1, letra2;  
    bool ans;

    while(getline(cin,cadena)){

        ans = true;
        i = 0;
        contador = 0;

        stack<string>pila;
        
        while(i<cadena.size() && ans == true){
            
            letra1 = cadena[i];

            if(letra1 == "("){

                pila.push(letra1);
            }
            
            else if(letra1 == "["){

                pila.push(letra1);
            }

            else if(letra1 == "{"){

                pila.push(letra1);
            }

            else if(letra1 == "<"){

                pila.push(letra1);
            }

            else if(letra1 == "*" && pila.top() == "("){
                
                pila.top()+="*";
                
            }

            if(letra1 == ")"){

                if(i>0){

                    letra2 = cadena[i-1];

                    if(!pila.empty() && pila.top() == "(*" && letra2 == "*"){
                        pila.pop();
                    }

                    else if(!pila.empty() && pila.top() == "(" && letra2 != "*"){
                        pila.pop();
                    }

                    else{
                        ans = false;
                    }
                }
                
                else{

                    if(!pila.empty() && pila.top() == "("){
                        pila.pop();
                    }

                    else{
                        ans = false;
                    }
                }
            }

            else if(letra1 == "]"){

                if(!pila.empty() && pila.top() == "["){
                    pila.pop();
                }

                else{
                    ans = false;
                }
            }

            else if(letra1 == "}"){

                if(!pila.empty() && pila.top() == "{"){
                    pila.pop();
                }

                else{
                    ans = false;
                }
            }

            else if(letra1 == ">"){

                if(!pila.empty() && pila.top() == "<"){
                    pila.pop();
                }

                else{
                    ans = false;
                }
            }

            i+=1;
            contador+=1;
        } 

        if(!pila.empty()){
            ans = false;
        }

        if(ans == true){
            printf("YES\n");
        }

        else{
            printf("NO %i\n", i);
        }
    }
    return 0;
}

File: online judge.kaner murillo/python/551.py
#
#551

from sys import stdin 
from collections import deque

cadena = stdin.readline().rstrip('\n')

while cadena != "":

    flag = True
    ans = True
    i = 0
    contador = 0
    corchete = ""
    pila = deque()

    while ans == True and i<len(cadena):

        flag = True

        if i < len(cadena)-1:

            if cadena[i] == "(" and cadena[i+1] == "*":
                pila.append("(*")
                flag = False
                i+=1

            elif(cadena[i] == "*" and cadena[i+1] == ")"):

                if len(pila) != 0 and pila[-1] == "(*":

                    pila.pop()
                    flag = False
                    i+=1
                
                else:
                    ans = False
                    i+=1

        if flag == True:

            if cadena[i] == "(" or cadena[i] == "[" or cadena[i] == "{" or cadena[i] == "<":

                pila.append(cadena[i])

            elif cadena[i] == ")" or cadena[i] == "]" or cadena[i] == "}" or cadena[i] == ">":
                
                if len(pila)!=0 and cadena[i] == ")" and pila[-1] == "(":

                    pila.pop()
                
                elif len(pila)!=0 and cadena[i] == "]" and pila[-1] == "[":
                    
                    pila.pop()
                
                elif len(pila)!=0 and cadena[i] == "}" and pila[-1] == "{":
                    
                    pila.pop()

                elif len(pila)!=0 and cadena[i] == ">" and pila[-1] == ">":
                    
                    pila.pop()

                else:
                    ans = False 

        i+=1
        contador+=1

    if len(pila)!= 0 and ans == True:
        ans = False
        contador+=1
    
    if ans == True:
        print("YES")
    
    else:
        print("NO",contador)

    cadena = stdin.readline().rstrip('\n')